SoD support

tstirrat opened the pull request:
Also includes the changes for Era: #8 (buff config changes) #7 (sort dropdowns) #10 (mobile sizing support) 

Work in progress changes:

- [ ] Druid 
   - [x] Bear - Updated Bear form coefficients, still need to check T1 6pc code
   - [x] Cat
   - [x] Moonkin - Added form and threat reductions, also some spell to form implications (Starsurge, Starfall)
   - [ ] Tree - WIP: Subtlety talent
- [x] Hunter - Added Ferocity (T1 2pc), Disengage, Feign Death, Distracting shot
- [ ] Mage
- [ ] Paladin - partial @DevilinAus ?
- [ ] Priest
- [x] Rogue
   - [x] Tank - Taunt. Main Gauche, T1 2pc, Unfair Advantage
- [ ] Shaman
- [x] Warrior
- [ ] Warlock

Voomlz commented on sod/spells.js:
```diff
@@ -1011,11 +1017,14 @@ const spellFunctions = {
+         9754: handler_modDamage(3.5, "Swipe (Rank 4)"),
+         9908: handler_modDamage(3.5, "Swipe"),
+
+       414644: handler_modDamage(3.5), // Lacerate
```
Lacerate from TBC has a specific handler, not sure if they change it for SOD ?

https://zidnae.gitlab.io/tbc-armor-penetration-calc/tbc_bear_tc.html

If it's the same you could consider taking function handler_lacerate from the TBC thing

tstirrat replied:
Lacerate in SoD appears to be a flat 3.5x for the hit and the dot. Checked in the disc, and I will do my own verification too

tstirrat commented on sod/spells.js:
```diff
@@ -99,11 +100,7 @@ const talents = {
+        return getThreatCoefficient((1.3 + 0.03 * rank) / 1.3); // additive, not multiplicative
 			}
 	
 		},
```
Ohh, that's how you do a per spell coeff. Thanks!

tstirrat commented:
@Voomlz I implemented the [threat coefficients for Rogue tank](https://github.com/Voomlz/voomlz.github.io/pull/6/commits/d2d6fb547b81ed474dfeb3ca1e763ec130c4310e) and noticed they were not applying. While debugging I've noticed that no players are receiving any buffs at all except for the implicit buffs like Salvation, Cat and Bear form etc.

essentially `spellId in buffs` checks all fail since `buffs` is empty. Therefore the rogue tank buff [Just a Flesh Wound](https://www.wowhead.com/classic/spell=400014/just-a-flesh-wound) does not apply, when it's [there in the logs](https://sod.warcraftlogs.com/reports/3kQ6nRY7W4fdtMJ8?fight=4&type=auras).

Have you seen this before or know what might be up? 

Check this log (adjust your local host): http://127.0.0.1:5500/sod/?id=https://sod.warcraftlogs.com/reports/3kQ6nRY7W4fdtMJ8

tstirrat commented:
> While debugging I've noticed that no players are receiving any buffs at all except for the implicit buffs like Salvation, Cat and Bear form etc.

Figured it out, any buffs that are not in `buffMultipliers` but we want to check elsewhere should be added manually into `notableBuffs` otherwise they get filtered out.

Voomlz commented:
I also had to add a lot of custom handlers and static variables to count stacks and what not in TBC
I wouldn't be surprised if a good chunk of the new features from SOD behave like some of those TBC spells
If that's the case I'd advise you to look at those implementations
Some are 'quick and dirty' fixes, but at least it gives correct threat values

tstirrat commented:
I did a big rework on the Rogue stuff so that we use buffImplications and initialBuffs instead of "talents".

Seems to work pretty good, the T1 2pc needs manual toggling, the other bits are inferred.

I raised a couple other PRs to Era that will benefit both Era and Sod, please give them a look. Once merged they will lessen the diff in this PR.

tstirrat commented:
> I wouldn't be surprised if a good chunk of the new features from SOD behave like some of those TBC spells

Yes and no. I did have a look. Many of the SoD runes are a single rank spell vs being many ranks in TBC (like Lifebloom and Nourish)

I'll keep looking at it for reference, of course

Voomlz commented:
Is there a lot of changes for warriors in SOD ?
If the main tanking class start to work, I think we could think about merging a first version with disclaimer and maybe gather some user feedback as well.

Are you on the discord btw?

tstirrat commented:
> I had this ugly T6 handling in TBC

Ah! I see that now.  I also was working on hacks in `era/` to look up `tables/buffs` to try work out initial buffs, and see you already did it. **We should replicate that into a common place that's used by all variants.**

> If the main tanking class start to work, I think we could think about merging a first version with disclaimer

Sounds good I'll fix up Pally RF and see what Warrior changes need to be done

tstirrat commented:
> Are you on the discord btw?

Yeah, I'm on most of the classic tank class discords, what's your name?

tstirrat commented:
I've pushed the Warrior updates, with optional Runes as buffs, and Verified the coefficients from the Disc:

Stances:
Defensive: 1.3, + Defiance 5/5 = 1.45
Gladiator: 0.7
Other stances: 0.8

Abilities:
Shield Bash: still 1.5x + flat threat
Shield Slam: 2.0x base - T2.5 6pc: another 2.0x
Revenge: 2.25
Devastate with Rune: 1.5x (might also have a base of 1.5x? unclear)
Thunder Clap with Rune: 1.5

In SoD, the tier bonuses are modelled in the game and come across as buffs ([example](https://www.wowhead.com/classic/spell=1214162/s03-item-taq-warrior-tank-4p-bonus)) in WCL (but not always consistent).. so I think it makes sense to model them as buffs in the tool that we can infer or turn on/off.

Let me know if that seems good enough to release

tstirrat commented:
Re other gear, I would like to pull across your gear check stuff, but allow the logic to be handled in the `<version>/spells.js` so we can differ it in each game version. There is a threat enchant to gloves, but I am pretty sure it differs in ID to TBC

And would love to bring across your initial buffs checking too, since the buffs table includes 100% uptime buffs like tier set bonuses and forms.

These bits can some later of course.

Voomlz commented:
> > Are you on the discord btw?
> 
> Yeah, I'm on most of the classic tank class discords, what's your name?

It's Vooml (without the z)

Voomlz commented:
> And would love to bring across your initial buffs checking too, since the buffs table includes 100% uptime buffs like tier set bonuses and forms.

That does sound like the T6 bonus on mangle typically ?

tstirrat commented:
> That does sound like the T6 bonus on mangle typically

Yeah and similar things.

Sod has some threat transfer items like Abandoned Wedding Ring (like vigilance, but from an item)

How are you feeling about merging since warrior is looking done?

Voomlz commented:
If you could just add a disclaimer in the SOD webpage stating that it's work in progress and feedback are welcome.

People used to bring good feedback with a couple of warcraft logs and twitch/youtube links for the in game threat value when things weren't correct as well. It can give you hint of things to fix.

BTW that's the discord server I was talking about (that is linked in the webpage https://discord.com/invite/3J4FGUNfW7)

Voomlz commented on sod/spells.js:
```diff
@@ -1,1163 +1,1528 @@
+         9754: handler_modDamage(3.5, "Swipe (Rank 4)"),
+         9908: handler_modDamage(3.5, "Swipe"),
+
+       414644: handler_modDamage(3.5), // Lacerate (Initial)
```
Are you sure of lacerate behaviour ?
For TBC the inital damage has a threat bonus, and then the dot has a multiplier

tstirrat replied:
Yeah, from the Disc [here](https://discord.com/channels/253205420225724416/1186591609819762750/1310758866841108480):

> Threat mods listed are correct.
> 
> Swipe has 0.1x AP scaling. 3.5x threat mod.
> Lacerate DoT has 0 AP scaling. 3.5x threat mod on initial hit and on the dot.
> Mangle has 1.0x Threat mod.

tstirrat commented on sod/spells.js:
```diff
@@ -909,6 +909,20 @@ function handler_threatOnHit(threatValue) {
     }
 }
 
+function handler_heroicStrikeThreatOnHit(threatValue) {
```
OK I'll remove it

tstirrat commented:
@Voomlz 

Ok, I've tried to have all the Warrior tank bits implemented. 

- Added T1/T2 core forged/T2.5 threat mods. 

Also:
- Added threat and subtlety (2%) gear enchants.
- Tried to get the Paladin threat mods implemented as best I can
   - There is some discrepancy with the HoR rune applying differently to un-talented RF that can probably be improved later
   - Ret paladins sim quite hight on threat, I assume this is because there is some buff/talent/rune missing that needs implementing.

I think it would be very helpful to see the coefficients split and labeled separately in the tooltip so they can be debugged better.

I might work on that as a next priority.
